File: Data Structures & Algorithms/validate-parentheses/submission-0.cpp
class Solution {
public:
    bool isValid(string s) {
        vector<int> stk;
        unordered_map<char, char> parenthesis {{')', '('}, {']', '['}, {'}', '{'}};
        for (char c : s) {
            if (c == '(' || c == '[' || c == '{')
                stk.push_back(c);
            if (c == ')' || c == ']' || c == '}') {
                if (stk.empty() || stk.back() != parenthesis[c])
                    return false;
                else
                    stk.pop_back();
            }
        }
        if (!stk.empty())
            return false;

        return true;
    }
};

File: Data Structures & Algorithms/validate-parentheses/submission-0.py
class Solution:
    def isValid(self, s: str) -> bool:
        pairs = {
            ")" : "(", 
            "]" : "[", 
            "}" : "{", 
        }

        stack = []

        for c in s:
            if c in ["(", "[", "{"]:
                stack.append(c)
            if c in pairs.keys():
                if pairs[c] != stack[-1]:
                    return False
                else:
                    stack.pop()
        
        if len(stack):
            return False

        return True